File: Part2_Regression/Section7_SVR/svr.R
# SVR

# Loading libraries
library(ggplot2)

# Importing the dataset
setwd("C:/Development/Courses/Kirill Eremenko Data Science Courses/Machine_Learning_A-Z/Part2_Regression/Section7_SVR")
dataset <- read.csv("Position_Salaries.csv")
dataset <- dataset[, 2:3]

# Splitting the dataset into Training set and Test set
# library(caTools)
# set.seed(123)  # like random_state
# split <- sample.split(dataset$Profit, SplitRatio=0.8)  # for training set
# training_set <- subset(dataset, split == TRUE)
# test_set <- subset(dataset, split == FALSE)

# Feature Scaling
# training_set[, 2:3] <- scale(training_set[, 2:3])  # Excluding categories since they're factors
# test_set[, 2:3] <- scale(test_set[, 2:3])

# Fitting SVR to the dataset
library(e1071)  # includes feature scaling
regressor <- svm(formula=Salary ~ ., data=dataset, type="eps-regression")

# Predicting a new result
y_pred <- predict(regressor, data.frame(Level=6.5))

# Visualising the SVR results
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary), color="red") +
  geom_line(aes(x=dataset$Level, y=predict(regressor, newdata=dataset)),
            color="blue") +
  ggtitle("Truth or Bluff (SVR)") +
  xlab("Level") +
  ylab("Salary")


# Visualising the SVR results
# Higher resolution and smoother curve
x_grid <- seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary), color="red") +
  geom_line(aes(x=x_grid, y=predict(regressor, newdata=data.frame(Level=x_grid))),
            color="blue") +
  ggtitle("Truth or Bluff (SVR)") +
  xlab("Level") +
  ylab("Salary")

File: Part2_Regression/Section7_SVR/svr.py
# SVR

# Data Preprocessing
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
from sklearn.preprocessing import StandardScaler # feature scaling
# from sklearn.cross_validation import train_test_split  # deprecated
# from sklearn.model_selection import train_test_split  # splitting the dataset
# from sklearn.linear_model import LinearRegression
# from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVR
# import statsmodels.formula.api as sm  # calculate p-values and other stats


# Importing the dataset
os.chdir("C:/Development/Courses/Kirill Eremenko Data Science Courses/\
Machine_Learning_A-Z/Part2_Regression/Section7_SVR")
dataset = pd.read_csv("Position_Salaries.csv")

# Create a matrix with all rows and the Level column only
X = dataset.iloc[:, 1:2].values

# Create the dependent variable vector; last column only
y = dataset.iloc[:, 2].values

# To see the full array, run the following
# np.set_printoptions(threshold=np.nan)

# Splitting the dataset into Training set and Test set - not creating here
# Only 10 rows so wouldn't provide value
"""
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                    random_state=0)
"""
# random_state set to 0 so we all get the same result
# 42 is a good choice for random_state otherwise

# Feature Scaling
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = sc_y.fit_transform(y)


# Fitting the SVR to the dataset
# Create your regressor here
regressor = SVR(kernel="rbf")
regressor.fit(X, y)

# Predicting a new result with SVR
# 2 pairs of brackets means an array of 1 line and 1 cell
y_pred = sc_y.inverse_transform(regressor.
                                predict(sc_X.transform(np.array([[6.5]]))))

# Visualising the SVR results
plt.scatter(X, y, color="red")
plt.plot(X, regressor.predict(X), color="blue")
plt.title("Truth or Bluff (SVR)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show()

# Visualising the SVR results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color="red")
plt.plot(X_grid, regressor.predict(X_grid), color="blue")
plt.title("Truth or Bluff (SVR)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show()
